create stencil for adhesive layers

wannessels opened the pull request:
- added --type option for adhesive, defaults to solderpaste
- changed zipping gerbers from subprocess.check_call("zip") to shutil.make_archive()
- fixed --cutout (layer names did not match)

yaqwsx commented on kikit/export.py:
```diff
@@ -1,21 +1,41 @@
 from pcbnew import *
 
+
+class LayerToPlot(Enum):
+    CuTop = (pcbnew.F_Cu, "Top layer")
+    CuBottom = (pcbnew.B_Cu, "Bottom layer")
+    PasteTop = (pcbnew.F_Paste, "Paste top")
+    PasteBottom = (pcbnew.B_Paste, "Paste bottom")
+    SilkTop = (pcbnew.F_SilkS, "Silk top")
+    SilkBottom = (pcbnew.B_SilkS, "Silk bottom")
+    MaskTop = (pcbnew.F_Mask, "Mask top")
+    MaskBottom = (pcbnew.B_Mask, "Mask bottom")
+    EdgeCuts = (pcbnew.Edge_Cuts, "Edges")
+    CmtUser = (pcbnew.Cmts_User, "V-CUT")
+    AdhesiveTop = (pcbnew.F_Adhes, "Adhesive top")
+    AdhesiveBottom = (pcbnew.B_Adhes, "Adhesive bottom")
+
+    def __init__(self, id: int, description: str):
+        self.id = id
+        self.description = description
+
```
I am unsure about this approach. I understand the need to provide "commonly used values"; however, this seems like an abuse of Enum a breaks type annotations.

If we want to allow the users to provide custom descriptions to the layers (as it was possible so far), this approach breaks the type system. The users will be forced to pass new tuples that, however, aren't part of the enum.

If we want to forbid it, I find a much nicer solution not to introduce new types and instead, internally have a function that maps layers to descriptions.

I understand the need for giving some type safety here - the code really needs it. But in that case, wouldn't a named tuple or a dataclass be a much better solution? Something like:
```
@dataclass
class LayerToPlot:
	layer: Layer
	description: str
```

And then, if required, provide some default values.

Or is there something I miss about using the enum?

wannessels replied:
IIRC I changed this because the cutout layer name was wrong. I didn't consider the custom descriptions. Dataclass looks like a good approach, I'll give that a try

yaqwsx commented on kikit/export.py:
```diff
@@ -71,62 +92,62 @@ def gerberImpl(boardfile, outputdir, plot_plan=fullGerberPlotPlan, drilling=True
-    popt.SetUseAuxOrigin(settings["UseAuxOrigin"])
-    popt.SetUseGerberX2format(False)
-    popt.SetDrillMarksType(0) # NO_DRILL_SHAPE
+    plot_controller = PLOT_CONTROLLER(board)
+    plot_options = plot_controller.GetPlotOptions()
+
+    plot_options.SetOutputDirectory(plotDir)
+
+    plot_options.SetPlotFrameRef(False)
+    plot_options.SetSketchPadLineWidth(FromMM(0.35))
+    plot_options.SetAutoScale(False)
+    plot_options.SetScale(1)
+    plot_options.SetMirror(False)
+    plot_options.SetUseGerberAttributes(False)
+    plot_options.SetIncludeGerberNetlistInfo(True)
+    plot_options.SetCreateGerberJobFile(True)
+    plot_options.SetUseGerberProtelExtensions(settings["UseGerberProtelExtensions"])
+    plot_options.SetExcludeEdgeLayer(settings["ExcludeEdgeLayer"])
+    plot_options.SetScale(1)
+    plot_options.SetUseAuxOrigin(settings["UseAuxOrigin"])
+    plot_options.SetUseGerberX2format(False)
+    plot_options.SetDrillMarksType(0) # NO_DRILL_SHAPE
```
I am not pleased about combining renaming and introducing new functionality in a single commit. It makes history unusable. If you want to trace a bug in history, you cannot easily tell what a behavior change was and what was only renaming.

If renaming is desirable, I prefer first to introduce a commit that changes the naming and then implement the new functionality (or in reverse order).

Note that if there is complete refactor of the function, that internally changes how the function operates then I consider it fine to rename and implement at the same time (as there is no meaningful mapping between the old and new function body).

However, I don't require you to change it; I will accept this PR it.

wannessels replied:
Agreed, renaming should've been a separate commit, I'll change it

yaqwsx reviewed (changes requested):
I am sorry it took me so long to review these changes. I left some of my notes in the comment, and please, let me know what you think.

wannessels commented on kikit/stencil.py:
```diff
@@ -420,21 +425,21 @@ def getComponents(board, references):
     return [f for f in board.GetFootprints() if f.GetReference() in references]
 
-def collectFootprintEdges(footprint, layerName):
+def collectFootprintEdges(footprint, layer: Layer):
```
👍 will do

wannessels commented on kikit/stencil_ui.py:
```diff
@@ -32,20 +53,14 @@ def createPrinted(**kwargs):
-    help="Comma separated list of components references to exclude from the stencil")
-@click.option("--cutout", type=str, default="",
-    help="Comma separated list of components references to cutout from the stencil based on the courtyard")
+@click.option("--jigsize", type=(int, int), default=(100, 100), help="Jig frame size in mm: <width> <height>")
+@click.option("--jigthickness", type=float, default=3, help="Jig thickness in mm")
+@click.option("--pcbthickness", type=float, default=1.6, help="PCB thickness in mm")
+@click.option("--registerborder", type=(float, float), default=(3, 1), help="Register borders in mm: <outer> <inner>")
+@click.option("--tolerance", type=float, default=0.05, help="Enlarges the register by the tolerance value")
+@click.option("--ignore", type=str, default="", help="Comma separated list of components references to exclude from the stencil")
+@click.option("--cutout", type=str, default="", help="Comma separated list of components references to cutout from the stencil based on the courtyard")
+@click.option("--type", type=EnumChoice(StencilType, case_sensitive=False), default="solderpaste", help="Stencil for SolderPaste or Adhesive")
```
Imho help is for printing on console, but less useful when reading the code. I find 1 parameter per line way more readable than trying to parse parameter name and help out of each line.
Same goes for 80 characters line length: that's a puch card relic, no need to keep it that short when we all have bigger monitors.

But that's just my opinion, happy to revert this if you want.

yaqwsx replied:
The 80-character limit isn't strict, and I often violate it if it makes the code more readable. But this is not the case. I also try to keep the code narrow, so you can easily fit 2-3 source codes side-by-side on a laptop.

Please, revert the formatting change.

wannessels commented:
I've made some changes:

- replaced enum with dataclass & constants
- reverted formatting change for click options help
- added typing to all changed methods
- set stencil_type default to SolderPaste, wherever the parameter was added, for backwards compatibility
